### PR TITLE
core: be nicer to read-only images

### PR DESCRIPTION
**Pull request by poettering**
Result of the discussions from #14442

@topimiettinen could you give this a whirl?

### CODE REVIEW

File: src/core/namespace.c [hunk 1]
```diff
@@ -1228,8 +1288,12 @@ int setup_namespace(
...
 
-                if (protect_system == PROTECT_SYSTEM_STRICT &&
-                    protect_home != PROTECT_HOME_NO &&
+                /* Make the whole image read-only if we can determine that we only access it in a read-only fashion. */
+                if (root_read_only(read_only_paths,
+                                   protect_system) &&
+                    home_read_only(read_only_paths, inaccessible_paths, empty_directories,
+                                   bind_mounts, n_bind_mounts, temporary_filesystems, n_temporary_filesystems,
+                                   protect_home) &&
                     strv_isempty(read_write_paths))
```
Comment (topimiettinen) [inline src/core/namespace.c:1288]: For completeness, this check for `read_write_paths` could also ignore the targets under `/home` or APIVFS.

Comment (poettering) [reply src/core/namespace.c:1288]: well, let's do that later, if anyone really cares ;-)

**Comment by topimiettinen** [comment]
Unfortunately something is wrong:
`Assertion 'IN_SET(open_flags, O_RDWR, O_RDONLY)' failed at src/shared/loop-util.c:28, function loop_device_make(). Aborting.`

File: src/shared/loop-util.c [hunk 2]
```diff
@@ -157,14 +158,29 @@ int loop_device_make(
 
 int loop_device_make_by_path(const char *path, int open_flags, uint32_t loop_flags, LoopDevice **ret) {
         _cleanup_close_ int fd = -1;
+        int r;
 
         assert(path);
         assert(ret);
-        assert(IN_SET(open_flags, O_RDWR, O_RDONLY));
+        assert(open_flags < 0 || IN_SET(open_flags, O_RDWR, O_RDONLY));
```
Comment (topimiettinen) [inline src/shared/loop-util.c:165]: The same change should be done for `loop_device_make()`, or make sure it never gets -1 for open_flags.

File: src/shared/loop-util.c [hunk 3]
```diff
@@ -157,14 +158,29 @@ int loop_device_make(
...
-                return -errno;
+        /* Passing < 0 as open_flags here means we'll try to open the device writable if we can, retrying
+         * read-only if we cannot. */
+
+        fd = open(path, O_CLOEXEC|O_NONBLOCK|O_NOCTTY|(open_flags >= 0 ? open_flags : O_RDWR));
+        if (fd < 0) {
+                r = -errno;
+
+                /* Retry read-only? */
+                if (open_flags >= 0 || !(ERRNO_IS_PRIVILEGE(r) || r == -EROFS))
```
Comment (topimiettinen) [inline src/shared/loop-util.c:175]: Why `r == -EROFS`? Shouldn't this be exactly the case to try with O_RDONLY?

Comment (poettering) [reply src/shared/loop-util.c:175]: yes, and we do. the check is negated, look closely!

**Comment by poettering** [comment]
force pushed a new version, ptal!

File: src/shared/loop-util.c (hunk 2 above)
Comment (poettering) [reply src/shared/loop-util.c:165]: nah, this function should patch open_flags to something valid if < 0. Fixed now.

Comment (topimiettinen) [reply src/shared/loop-util.c:165]: I can confirm that the assertion failure is gone after changing also `loop_device_make()` the same way.

File: src/core/namespace.c [hunk 4]
```diff
@@ -1190,6 +1190,66 @@ static void normalize_mounts(const char *root_directory, MountEntry *mounts, siz
...
+         * writable one from a source outside of /home. */
+
+        for (i = 0; i < n_bind_mounts; i++) {
+                if (!path_equal(bind_mounts[i].destination, "/home"))
+                        continue;
+
+                if (bind_mounts[i].read_only)
+                        return true;
+
+                if (!path_startswith(bind_mounts[i].source, "/home"))
```
Comment (topimiettinen) [inline src/core/namespace.c]: I'm doing `BindPaths=/home:/home`, so the root is still read-write even if /home is a different partition. How about looking at the block devices for /home and root. If they are different, mounting root as read-only will not affect /home?

**Comment by topimiettinen** [comment]
The assertion is gone, but root is still mounted read-write.

File: src/core/namespace.c (hunk 4 above)
Comment (poettering) [reply src/core/namespace.c]: ah, eh, i forgot that BindPaths= only binds stuff from the host, never from the container... hence it's sufficient to check the destination, the source doesn't need to be checked and read-only or not doesn't need to be checked either. will prep fix.

**Comment by poettering** [comment]
force pushed a new version. does this one work for you @topimiettinen?

**Comment by topimiettinen** [comment]
Great, root is mounted read-only!

**Comment by anitazha** [comment]
The PR LGTM. However I noticed there are no tests for `RootImage=` and many of the options under https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Paths. What's the plan here for that?

**Comment by poettering** [comment]
@anitazha testing RootImage= is hard since our tests run inside OS images built for this test during the test build, and it's nasty to build test images to include in test images to test images inside of the test images ;-)

I mean, it would be excellent to test RootImage= more... much like a lot of other stuff, actually ;-)

Long story short: we'd always be delighted to have more tests for that, but so far we stuck our head in the sand for such harder to test stuff and ran without it...

**Comment by poettering** [comment]
Anyway, two people liked this, and of them tested this. let's merge! thanks for the reviews!

**Pull request merged**